### PR TITLE
Spec for superfluid messages - WIP

### PR DESCRIPTION
**Pull request by UnityChaos**
Closes: #XXX

## Description

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/gaia/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: x/superfluid/keeper/twap_price.go [hunk 1]
```diff
@@ -37,6 +37,10 @@ func (k Keeper) DeleteEpochOsmoEquivalentTWAP(ctx sdk.Context, denom string) {
 	prefixStore.Delete([]byte(denom))
 }
 
+// Not clear to me how this works?
+// looks like it expects twap to be already in state?
+// but we're supposed to be using spot price instead in MVP?
+// also wouldn't this depend on the ordering of the assets? Osmo/X vs X/Osmo?
 func (k Keeper) GetEpochOsmoEquivalentTWAP(ctx sdk.Context, denom string) sdk.Dec {
```
Comment (ValarDragon) [inline x/superfluid/keeper/twap_price.go:44]: Its set in hooks.go, which contains the epoch code.

File: x/superfluid/spec/03_messages.md [hunk 2]
```diff
@@ -3,3 +3,45 @@ order: 3
...
+- Get the `IntermediaryAccount` for this `lock` `Denom` and `ValAddr` pair.
+- Mint `Osmo` to match amount in `lock` and send to `IntermediaryAccount`
+- Create a delegation from `IntermediaryAccount` to `Validator`
+- Create a new perpetual `Gauge` for distributing staking payouts to locks of a synethic asset based on this `Validator` / `Denom` pair.
+- Create a connection between this `lockID` and this `IntermediaryAccount`
+
+### Questions
+
+- Where are we checking that the lock is of an LP share?
+- Where are we checking that the pool contains Osmo?
```
Comment (ValarDragon) [inline x/superfluid/spec/03_messages.md:35]: In superfluid asset creation

File: x/superfluid/spec/03_messages.md [hunk 3]
```diff
@@ -3,3 +3,45 @@ order: 3
...
+- Where are we checking that this pool is approved for superluid?
```
Comment (ValarDragon) [inline x/superfluid/spec/03_messages.md:36]: The approval process is encoded by the superfluid assets approval

File: x/superfluid/keeper/twap_price.go (hunk 1 above)
Comment (ValarDragon) [reply x/superfluid/keeper/twap_price.go:44]: The function used to calculate it is above in this file, at calculateOsmoBackingPerShare

File: x/superfluid/keeper/stake.go [hunk 4]
```diff
@@ -200,6 +200,7 @@ func (k Keeper) SuperfluidDelegate(ctx sdk.Context, sender string, lockID uint64
 
 	// check unbonding synthetic lockup already exists on this validator
 	// in this case automatic superfluid undelegation should fail and it is the source of chain halt
+	// TODO/FIXME ... ^ what?
```
Comment (ValarDragon) [inline x/superfluid/keeper/stake.go:203]: 👀

File: x/superfluid/keeper/twap_price.go (hunk 1 above)
Comment (UnityChaos) [reply x/superfluid/keeper/twap_price.go:44]: :+1: I see, so by the time someone can send a delegation tx, the twap (spot) value already exists in state. And if it doesn't exist, that indicates that either:
The pool isn't approved for superfluid, the pool doesn't have osmo in it, or the asset isn't an LP token.

**Review by ValarDragon — APPROVED** [review]

**Comment by ValarDragon** [comment]
Merging this to avoid branch conflicts

**Pull request merged**